### PR TITLE
General superior animal mob optimization

### PR DESCRIPTION
**Pull request by ma44**
## About The Pull Request
Makes superior animals (mainly roaches) more cost friendly for the server CPU to handle; this could be greatly improved should the target scope be changed (all mobs => only humans) and/or porting of "search through list of mobs by zlevel", but this can open up some discussion on the matter.

## Why It's Good For The Game
Less CPU consumption for the same effect of mobs killing each other and players; more GLOBs for coders to play around with.

## Changelog
:cl: ma44
tweak: Roaches and other superior animals are more optimized and shouldn't cause as much lag as before.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/superior_animal/ai.dm
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/superior_animal/ai.dm:10]: What proof do you have that this is more performant?

I can imagine it'll scale poorly with having a lot of mobs.

File: code/modules/mob/living/carbon/superior_animal/roach/life.dm
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/superior_animal/roach/life.dm:22]: The istype checks are redundant with your changes

File: code/modules/mob/living/carbon/superior_animal/roach/roach.dm
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/superior_animal/roach/roach.dm:43]: Undeclared change.

File: code/modules/mob/living/carbon/superior_animal/roach/types/support.dm
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/superior_animal/roach/types/support.dm:47]: Rather than the target mob check you have made above, have it here as `
if(target_mob && prob(7))`, otherwise you're skipping the blattedin regeneration.

File: code/modules/mob/living/carbon/superior_animal/ai.dm
Comment (MadmanMartian) [inline code/modules/mob/living/carbon/superior_animal/ai.dm:42]: This is much different to hearers, but should be cheaper.

File: code/modules/mob/living/carbon/superior_animal/ai.dm
Comment (TheShown911) [reply code/modules/mob/living/carbon/superior_animal/ai.dm:10]: Hearers is very heavy on the CPU, but i know that big lists can have an impact on memory and may cause OOR so i would like to see some benchmarks of this before approval, and those benchmarks should not include the changes from ` code/modules/mob/living/living_defines.dm` lines 11 and 12.

**Comment by TorinoFermic** [comment]
Changes given by MadMan should be completed

**Comment by ma44** [comment]
Will be able to finish reworking this in 1-2 days from now.

**Comment by drexample** [comment]
@ma44 Conflict

File: code/game/atoms_movable.dm
Comment (Evankhell561) [inline code/game/atoms_movable.dm:393]: for(var/atom/movable/AM in src)

File: code/modules/mob/living/carbon/superior_animal/giant_spider/types/nurse.dm
Comment (Evankhell561) [inline code/modules/mob/living/carbon/superior_animal/giant_spider/types/nurse.dm:49]: if(C.stat != CONSCIOUS)

File: code/modules/mob/mob_defines.dm
Comment (Evankhell561) [inline code/modules/mob/mob_defines.dm:201]: var/registered_z

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/maps/swapmaps.dm
Comment (Evankhell561) [inline code/modules/maps/swapmaps.dm:328]: you are applying different logic to the original code and I am not sure that is your intention or that it is a good idea.

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

**Review by Hopekz — APPROVED** [review]
Good to go here once this makes it past code review.
nhp8v4DVda_nice-export

File: code/controllers/subsystems/processing/mobs.dm
Comment (Evankhell561) [inline code/controllers/subsystems/processing/mobs.dm:19]: What is the second oargument on the list for?

File: code/game/atoms_movable.dm
Comment (Evankhell561) [inline code/game/atoms_movable.dm:392]: fine, but... why send this signal if no one is receiving it?

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm
Comment (Evankhell561) [inline code/game/atoms_movable.dm:391]: and if nobody receives this signal ... what is this proc for?

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/__HELPERS/global_lists.dm
Comment (Evankhell561) [inline code/__HELPERS/global_lists.dm:16]: fine, but what do you want this list for if you're not going to use it?

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/life.dm
Comment (Evankhell561) [inline code/modules/mob/living/life.dm:75]: if(registered_z != T.z)

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/superior_animal/ai.dm
Comment (ma44) [reply code/modules/mob/living/carbon/superior_animal/ai.dm]: With the new code, the amount of mobs present in each zlevel should be overall less things to look through compared to hearers() which returns a grid of ((viewRange * 2) + 1) x ((viewRange * 2) + 1) around src and all of it's contents.

Comment (ma44) [reply code/modules/mob/living/carbon/superior_animal/ai.dm]: Outdated code, it currently utilizes the following
`for(var/atom/thing in SSmobs.mob_living_by_zlevel[((get_turf(src)).z)])`

File: code/__HELPERS/global_lists.dm
Comment (ma44) [reply code/__HELPERS/global_lists.dm:16]: Leftover from attempted implementation, no longer needed for this PR but it could stay if someone wishes to do something with it.

File: code/controllers/subsystems/processing/mobs.dm
Comment (ma44) [reply code/controllers/subsystems/processing/mobs.dm:19]: Creates a multidimensional list consisting of `world.maxz` lists, each containing 0 elements

File: code/modules/mob/living/carbon/superior_animal/roach/types/support.dm
Comment (Evankhell561) [inline code/modules/mob/living/carbon/superior_animal/roach/types/support.dm:46]: I wonder if the proc life of cockroaches and of all superior animals continues to run after death, and that is the biggest problem. I ask because in this code I do not see that verification.

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/superior_animal/roach/types/support.dm
Comment (ma44) [reply code/modules/mob/living/carbon/superior_animal/roach/types/support.dm:46]: From what I can tell through the power of var editing, `Life()` does indeed stop ticking once the roach has been `Death()`'d, so I suppose this won't need a "if alive" check.

Comment (Evankhell561) [reply code/modules/mob/living/carbon/superior_animal/roach/types/support.dm:46]: don't worry, I studied the life code better and everything seems fine.

**Review by TorinoFermic — APPROVED** [review]

File: code/game/atoms_movable.dm
Comment (Evankhell561) [inline code/game/atoms_movable.dm:404]:
```suggestion
	for (var/atom/movable/AM in src) // Notify contents of Z-transition. This can be overridden IF we know the items contents do not care.
```

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm
Comment (Evankhell561) [inline code/game/atoms_movable.dm:404]:
```suggestion
/atom/movable/proc/onTransitZ(old_z, new_z)//uncomment when something is receiving this signal
	/*SEND_SIGNAL(src, COMSIG_MOVABLE_Z_CHANGED, old_z, new_z)
	for(var/atom/movable/AM in src) // Notify contents of Z-transition. This can be overridden IF we know the items contents do not care.
		AM.onTransitZ(old_z,new_z)*/ 
```

**Review by Evankhell561 — DISMISSED** [review]

**Pull request merged**
